contracts/Cryptoxo.sol

```diff
@@ -0,0 +1,254 @@
+    event ForciblyDraw(string);
+    event ReturnToInintialStateComplete(string);
+    event UserWin(address _winner, uint _amountWon, string _compareMethodName);
+    event GameIsDraw(string _drawGameMsg, uint _amountToPayback, string _goesTo, address _playerX, address _PlayerO);
+    event MoveMade(address _currentPlayer, string _playerVal, uint8 _positionColumn, uint8 _positionRow);
+    event MatchContinius(string _matchContinue);
+    event showOxoBoardString(
+                            string _lineNum, 
+                            string _firstPosition, 
+                            string _secondPosition, 
+                            string _thirdPosition
+                            );
+                            
+    // modifier for function to return ETH in case 
+    // playerRateProposer made an offer
+    // but none accepted the chellenge
+    modifier onlyPlayerRateProposer() {
+        require(msg.sender == playerRateProposer);
+        _;
+    }
+    
+    // modifier to use forciblyDraw() function 
+    // in case something goes wrong and one of players disappears
+    // to manually draw the game
+    modifier onlyOwner() {
+        require(msg.sender == owner);
+        _;
+    }
+    
+    // according to rules onwer can't play the game
+    // owner can only be a reffery 
+    // in case one of players disappears
+    // to manually draw the game
+    modifier onlyNotOwner() {
+        require(msg.sender != owner);
+        _;
+    }
+    
+    
+    
+    // init
+    function Cryptoxo(){
+        // adding owner in case a contract needs a suicide
+        owner = msg.sender;
+        //movesCounter = 0;
+        
+        // creating an empty oxo board;
+        oxoBoard[0] = ["-", "-", "-"];
+        oxoBoard[1] = ["-", "-", "-"];
+        oxoBoard[2] = ["-", "-", "-"];
+    }
+    
+    
+    
+    
+    function offerRate() public payable onlyNotOwner returns (uint){
+        require(this.balance == msg.value);
+        
+        rateProposed = msg.value;
+        playerRateProposer = msg.sender;
+        
+        return rateProposed;
+    }
+    
+    function replyOffer() public payable onlyNotOwner returns (bool){
+        require(rateProposed != 0);
+        require(msg.sender != playerRateProposer);
+        require(msg.value == rateProposed);
+       
+        players.x = playerRateProposer;
+        players.o = msg.sender;
+        playersVal[playerRateProposer] = "X";
+        playersVal[msg.sender] = "O";
+        playerRateProposer = 0;
+        rateProposed = 0;
+        
+        return true;
+    }
+    
+    function seeOfferRate() public returns (uint){
+        require(rateProposed != 0);
+        return rateProposed;
+    }
+    
+    function makeMove(uint8 _column, uint8 _row) public returns (bool){
+        require(keccak256(playersVal[msg.sender]) != keccak256(0));
+        require(msg.sender != previousMovePlayer);
+        require(keccak256(oxoBoard[_row-1][_column-1]) == keccak256('-'));
```

Comment:
why row - 1 and column - 1? 
If I enter 0, 0 than I'll have revert

Reply:
yes, you will have revert. 'row - 1' is done to be more human stile for a player (usual humans don't know that array stars from zero but not from one)

Reply:
meh, I forgot that common humans exist